### PR TITLE
Fix more invalidations from overloading `==`

### PR DESCRIPTION
**Pull request by timholy**
Other than a PR to Pkg (which needs #36280), this is pretty much everything that I think can be done on #36005 without changes to the compiler itself. The combination of #36255, #36280, #36281, the upcoming Pkg PR, some PRs to JuliaInterpreter/LoweredCodeUtils/Revise, and this PR is to reduce the total count of (non-unique) invalidations from more than 3k to 529. Of those, 281 are from `==(::Symbol, ::Any)`, which I do not see a reasonable way to address without changes to the compiler itself (some details were mentioned in #36255).

### CODE REVIEW

**Comment by timholy** [comment]
BTW, we are making progress but there's still a lot of work ahead. Numbers in the left two columns are where we were in May (around the time of the 1.5 feature freeze) when I was actively working on the blog post, numbers in the right two columns are where we are now on this PR and all the ones referenced above:

| Package | more specific (May) | ambiguous (May) | more specific (now) | ambiguous (now) |
|:------- | ------------------:| --------:| -----:| -----:|
| Example | 0 | 0  | 0  | 0  |
| Revise | 7 |  0 | 0 | 0 |
| FixedPointNumbers | 170 | 387 | 34 | 470 |
| SIMD | 3903 | 187 | 5  | 874 |
| StaticArrays | 989 | 3133 | 593 | 2060 |
| Optim | 1643 | 2921 | 760 | 2103 |
| Images | 1749 | 3671 | 610 | 2515 |
| Flux | 1991 | 3460 | not loadable (CUDAnative) |  |
| Plots | 1542 | 4302 | 1268 | 3028 |
| DataFrames | 4919 | 783 | 3520 | 1833 |
| JuMP | 2145 | 4670 | not loadable (MutableArithmetics) |  |
| Makie | 6233 | 5526 | 710 | 4081 |
| DifferentialEquations | 5152 | 6218 | 3787 | 3432 |

Obviously, the single biggest advance will come when the ambiguous cases go away, but #36200 and all these PRs to improve inferrability are starting to have a pretty big cumulative effect. @SimonDanisch and @eschnett should be particularly pleased.

I'm not quite sure why the ambiguity numbers have changed so much, but either it's changes in the packages (I wasn't careful to insist on the same versions) or the changes in inference have changed the dependency trees in a way that affects the ambiguity cases, too. (That actually makes sense.)

File: base/cartesian.jl [hunk 1]
```diff
@@ -353,17 +353,27 @@ const exprresolve_cond_dict = Dict{Symbol,Function}(:(==) => ==,
     :(<) => <, :(>) => >, :(<=) => <=, :(>=) => >=)
 
 function exprresolve_arith(ex::Expr)
-    if ex.head === :call && haskey(exprresolve_arith_dict, ex.args[1]) && all([isa(ex.args[i], Number) for i = 2:length(ex.args)])
-        return true, exprresolve_arith_dict[ex.args[1]](ex.args[2:end]...)
+    if ex.head === :call
+        callee = ex.args[1]
+        if isa(callee, Symbol)
+            if haskey(exprresolve_arith_dict, callee) && all([isa(ex.args[i], Number) for i = 2:length(ex.args)])
```
Comment (JeffBezanson) [inline base/cartesian.jl:359]:
```suggestion
            if haskey(exprresolve_arith_dict, callee) && all(Bool[isa(ex.args[i], Number) for i = 2:length(ex.args)])
```

File: stdlib/REPL/src/REPLCompletions.jl [hunk 2]
```diff
@@ -626,7 +628,7 @@ function completions(string, pos, context_module=Main)::Completions
         if isa(ex, Expr)
             if ex.head==:call
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPLCompletions.jl:629]:
```suggestion
            if ex.head === :call
```

File: stdlib/REPL/src/REPLCompletions.jl [hunk 3]
```diff
@@ -626,7 +628,7 @@ function completions(string, pos, context_module=Main)::Completions
...
                 return complete_methods(ex, context_module), first(frange):method_name_end, false
-            elseif ex.head==:. && ex.args[2] isa Expr && ex.args[2].head==:tuple
+            elseif ex.head==:. && ex.args[2] isa Expr && (ex.args[2]::Expr).head==:tuple
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPLCompletions.jl:631]:
```suggestion
            elseif ex.head === :. && ex.args[2] isa Expr && (ex.args[2]::Expr).head === :tuple
```

**Comment by timholy** [comment]
Wow, that `downloads` test fails a lot...

**Pull request merged**